### PR TITLE
samples: nrf9160: modem_shell: add modem_trace commands

### PR DESCRIPTION
**Pull request by gregersrygg**
The modem_trace command can be used to start/stop modem tracing.
Backends that allow storing of trace data can also print the size of
stored trace data, send trace data to the cloud (only Memfault for now)
and clear the stored trace data.

**This PR depends on #10199**
(Please ignore that commit when reviewing this PR)

### CODE REVIEW

File: modules/memfault-firmware-sdk/memfault_integration.c [hunk 1]
```diff
@@ -185,4 +186,13 @@ int memfault_ncs_device_id_set(const char *device_id, size_t len)
 	return 0;
 }
 
+#if defined(CONFIG_NRF_MODEM_LIB)
+NRF_MODEM_LIB_ON_INIT(memfault_ncs_init_hook, on_modem_lib_init, NULL);
+static void on_modem_lib_init(int ret, void *ctx)
```
Comment (jtguggedal) [inline modules/memfault-firmware-sdk/memfault_integration.c:191]: Missing blank line before function

File: samples/nrf9160/memfault/overlay-modem-trace.conf [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Modem trace
```
Comment (jtguggedal) [inline samples/nrf9160/memfault/overlay-modem-trace.conf:1]: Missing license

File: samples/nrf9160/memfault/overlay-modem-trace.conf [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+CONFIG_NRF_MODEM_LIB_TRACE=y
+CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_FLASH=y
+CONFIG_NRF_MODEM_LIB_SYS_INIT=n
+
+#increase trace write speed
+CONFIG_FCB_ALLOW_FIXED_ENDMARKER=y
+
+#CONFIG_NRF_MODEM_TRACE_FLASH_NOSPACE_ERASE_OLDEST=y
```
Comment (jtguggedal) [inline samples/nrf9160/memfault/overlay-modem-trace.conf:9]: Please remove if unused, same for others

File: samples/nrf9160/modem_shell/config/memfault_platform_config.h [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#pragma once
```
Comment (jtguggedal) [inline samples/nrf9160/modem_shell/config/memfault_platform_config.h:1]: Missing license

File: samples/nrf9160/modem_shell/config/memfault_platform_config.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+/*
+ * Platform overrides for the default configuration settings in the
+ * memfault-firmware-sdk. Default configuration settings can be found in
+ * "<NCS folder>/modules/lib/memfault-firmware-sdk/components/include/memfault/default_config.h"
+ */
+
+/* Uncomment the definition below to override the default setting for
```
Comment (jtguggedal) [inline samples/nrf9160/modem_shell/config/memfault_platform_config.h:9]: Copy/paste leftovers, can be removed

File: samples/nrf9160/modem_shell/CMakeLists.txt [hunk 6]
```diff
@@ -33,8 +33,11 @@ add_subdirectory_ifdef(CONFIG_MOSH_PPP src/ppp)
 add_subdirectory_ifdef(CONFIG_MOSH_FOTA src/fota)
 add_subdirectory_ifdef(CONFIG_MOSH_REST src/rest)
 add_subdirectory_ifdef(CONFIG_MOSH_STARTUP_CMDS src/startup_cmd)
+add_subdirectory_ifdef(CONFIG_NRF_MODEM_LIB_TRACE src/modem_trace)
 add_subdirectory(src/cloud)
 
+zephyr_include_directories(config)
```
Comment (jtguggedal) [inline samples/nrf9160/modem_shell/CMakeLists.txt:39]: This should probably be a conditional include

File: samples/nrf9160/modem_shell/overlay-memfault.conf [hunk 7]
```diff
@@ -0,0 +1,21 @@
+#
+# Copyright (c) 2023 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+# Memfault
+CONFIG_MEMFAULT=y
+CONFIG_MEMFAULT_NCS_PROJECT_KEY="OxWB2QrqNxdKQTTGKnJkwehqCdrMtZCE"
```
Comment (jtguggedal) [inline samples/nrf9160/modem_shell/overlay-memfault.conf:9]: Should be removed (content, the config can stay as empty or with dummy value)

File: samples/nrf9160/modem_shell/overlay-modem-trace.conf [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+# Copyright (c) 2023 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+CONFIG_NRF_MODEM_LIB_TRACE=y
+CONFIG_NRF_MODEM_LIB_SYS_INIT=n
+CONFIG_NRF_MODEM_LIB_TRACE_LEVEL_OFF=y
+
+# modem trace uart backend
```
Comment (jtguggedal) [inline samples/nrf9160/modem_shell/overlay-modem-trace.conf:11]: Nit: Preferable to start comments with uppercase

File: samples/nrf9160/modem_shell/src/main.c [hunk 9]
```diff
@@ -257,7 +257,7 @@ void main(void)
 
 #if !defined(CONFIG_LWM2M_CARRIER)
 	/* Get Modem library initialization return value. */
-	err = modem_lib_init_result;
+	err = nrf_modem_lib_init(NORMAL_MODE);
```
Comment (jtguggedal) [inline samples/nrf9160/modem_shell/src/main.c:260]: Is this change intentional?

File: samples/nrf9160/modem_shell/src/modem_trace/modem_trace_memfault.c [hunk 10]
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include <stdint.h>
```
Comment (jtguggedal) [inline samples/nrf9160/modem_shell/src/modem_trace/modem_trace_memfault.c:6]: Nit: Missing blank line

File: samples/nrf9160/modem_shell/src/modem_trace/modem_trace_memfault.c [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+#include <modem/nrf_modem_lib_trace.h>
+
+#include "modem_trace_memfault.h"
+#include "mosh_print.h"
+
+static bool memfault_cdr_source_registered;
+static const char *mimetypes = { MEMFAULT_CDR_BINARY };
+static sMemfaultCdrMetadata trace_recording_metadata = {
+	.start_time.type = kMemfaultCurrentTimeType_Unknown,
+	.mimetypes = &mimetypes,
```
Comment (jtguggedal) [inline samples/nrf9160/modem_shell/src/modem_trace/modem_trace_memfault.c:21]: This will work as long as the number of `.num_mimetypes`  is 1

File: samples/nrf9160/modem_shell/src/modem_trace/modem_trace_memfault.c [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+	.num_mimetypes = 1,
+	.data_size_bytes = 0,
+	.collection_reason = "modem traces",
+};
+
+static bool has_modem_traces;
+static bool has_cdr_cb(sMemfaultCdrMetadata *metadata)
```
Comment (jtguggedal) [inline samples/nrf9160/modem_shell/src/modem_trace/modem_trace_memfault.c:28]: Missing blank line before function

File: samples/nrf9160/modem_shell/src/modem_trace/modem_trace_memfault.h [hunk 13]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#ifndef MOSH_MODEM_TRACE_MEMFAULT_H
```
Comment (jtguggedal) [inline samples/nrf9160/modem_shell/src/modem_trace/modem_trace_memfault.h:6]: Missing blank line

**Review by jtguggedal — COMMENTED** [review]
Nice work! Did a quick pass, mostly minor things

File: samples/nrf9160/memfault/overlay-modem-trace.conf (hunk 2 above)
Comment (gregersrygg) [reply samples/nrf9160/memfault/overlay-modem-trace.conf:1]: Oh, this file wasn't supposed to be here. Will remove it.

File: samples/nrf9160/modem_shell/src/main.c (hunk 9 above)
Comment (gregersrygg) [reply samples/nrf9160/modem_shell/src/main.c:260]: I probably need to improve this logic. We need to call `nrf_modem_lib_init` manually when modem tracing is enabled. Perhaps best to check if `CONFIG_NRF_MODEM_LIB_SYS_INIT` is disabled?

Comment (gregersrygg) [reply samples/nrf9160/modem_shell/src/main.c:260]: Instead of checking `CONFIG_NRF_MODEM_LIB_SYS_INIT​` in main and have a code branch. How about I just move ​`CONFIG_NRF_MODEM_LIB_SYS_INIT=n`​ from `​overlay-modem-trace.conf`​ to `​prj.conf`​? So we always initialize the modem library manually (unless CONFIG_LWM2M_CARRIER).

File: samples/nrf9160/modem_shell/src/modem_trace/modem_trace_memfault.c (hunk 11 above)
Comment (gregersrygg) [reply samples/nrf9160/modem_shell/src/modem_trace/modem_trace_memfault.c:21]: I reverted back to how it was originally in the Memfault docs.

File: samples/nrf9160/modem_shell/src/main.c (hunk 9 above)
Comment (gregersrygg) [reply samples/nrf9160/modem_shell/src/main.c]: After thinking more about it, I think it might be unclear why `CONFIG_NRF_MODEM_LIB_SYS_INIT` would be disabled in prj.conf. So landed on keeping the configs as is and having a check in `main.c` instead.

File: samples/nrf9160/modem_shell/src/modem_trace/modem_trace_shell.c [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+{
+	enum nrf_modem_lib_trace_level level = (enum nrf_modem_lib_trace_level)data;
+	return modem_traces_start(level);
+}
+
+void nrf_modem_lib_trace_callback(enum nrf_modem_lib_trace_event evt)
+{
+	int err;
+
+	mosh_error("Error from nrf_modem_lib_trace: %d", evt);
```
[the remark this answers is not in the document]
Comment (lemrey) [reply samples/nrf9160/modem_shell/src/modem_trace/modem_trace_shell.c:95]: I agree, this is not an error condition; it is simply a callback for an event that signals the application that the FLASH is full. No error has occurred.

[the remark this answers is not in the document]
Comment (tokangas) [reply samples/nrf9160/modem_shell/src/modem_trace/modem_trace_shell.c:95]: I think it could be a warning to make it a bit more visible, otherwise it might be easy to miss it if there's a lot of other output. Even though it's something that is expected to happen, it's still better that the user notices it and does not assume that tracing is still active.

File: samples/nrf9160/modem_shell/src/modem_trace/modem_trace_shell.c [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+		}
+	}
+}
+
+void send_traces_to_memfault(void)
+{
+#ifdef CONFIG_MEMFAULT
+	uint32_t duration_ms = trace_duration_ms;
+
+	if (duration_ms == 0) { /* modem tracing active */
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/modem_trace/modem_trace_shell.c:111]: It could also be that tracing has not been started? Would it be better to also check that `trace_start_time_ms` is not 0?

**Review by tokangas — COMMENTED** [review]
Nice, not much to comment.

File: samples/nrf9160/modem_shell/src/modem_trace/modem_trace_shell.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (gregersrygg) [reply samples/nrf9160/modem_shell/src/modem_trace/modem_trace_shell.c:95]: Thanks. I agree this shouldn't be an error. Will change the message of the first `mosh_error` and reduce it to `mosh_print`. Will keep the message when flash is full and reduce it to a `mosh_warn`.

File: samples/nrf9160/modem_shell/src/modem_trace/modem_trace_shell.c [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+		"Stop modem tracing.", modem_traces_stop),
+	SHELL_CMD(clear, NULL,
+		"Clear captured trace data and prepare the backend for capturing new traces.\n"
+		"This operation is only supported with some trace backends.", modem_traces_clear),
+	SHELL_CMD(size, NULL,
+		"Read out the size of stored modem traces.\n"
+		"This operation is only supported with some trace backends.", modem_trace_size),
+	SHELL_CMD(send, &send_cmd,
+		"Collection of subcommands for handling stored trace data.\n"
+		"The stored traces are automatically deleted after send.", NULL),
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/modem_trace/modem_trace_shell.c:152]: Are they also deleted if sending fails?

**Review by trantanen — COMMENTED** [review]
Very nice feature. Thanks a lot for the effort!

File: samples/nrf9160/modem_shell/README.rst [hunk 17]
```diff
@@ -453,6 +453,69 @@ Examples
...
+----
+
+Modem traces
+============
+
+MoSh command: ``modem_traces``
+
+You can use the modem_trace commands for controlling the trace functionality in the modem.
+
+See :ref:`nrf_modem_lib_readme` (should link to https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/modem/nrf_modem_lib.html#modem-trace-module) for information on how to configure modem tracing and the built in
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply samples/nrf9160/modem_shell/README.rst:466]: We already have the label `modem_trace_module` in `nrf_modem_lib.rst` that points to this section.

[the remark this answers is not in the document]
Comment (gregersrygg) [reply samples/nrf9160/modem_shell/README.rst:466]: This is just a rough draft. I'm not good at RST, so will get help from @divipillai 😊

File: samples/nrf9160/modem_shell/src/modem_trace/modem_trace_shell.c (hunk 16 above)
Comment (gregersrygg) [reply samples/nrf9160/modem_shell/src/modem_trace/modem_trace_shell.c:152]: Unfortunately, yes. I have discussed with @eivindj-nordic that the trace read function has an argument to indicate if you want it to delete after read or not, but it was too close to RC1 to start implementing that. So it's a future improvement.

Comment (trantanen) [reply samples/nrf9160/modem_shell/src/modem_trace/modem_trace_shell.c:152]: Understood

File: samples/nrf9160/modem_shell/README.rst [hunk 18]
```diff
@@ -453,6 +453,69 @@ Examples
...
+cloud. The flash backend can store modem traces to the external flash on the nRF9160 DK and read
+them back later for uploading.
+
+To enable modem traces with flash backend, you can use :file:`overlay-modem-trace.conf`. This also
+requires a device tree overlay for the external flash (:file:`nrf9160dk_ext_flash.overlay`).
+
+Send to Memfault
+----------------
+
+https://developer.nordicsemi.com/nRF_Connect_SDK/doc/2.2.0/nrf/libraries/others/memfault_ncs.html#using-memfault-sdk
```
[the remark this answers is not in the document]
Comment (divipillai) [reply samples/nrf9160/modem_shell/README.rst]: Added links to both memfault and Memfault SDK.

File: samples/nrf9160/modem_shell/README.rst [hunk 19]
```diff
@@ -453,6 +453,69 @@ Examples
...
+See (link above) for how to register an account and get the project key.
+The project key should be configured with :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY`.
+The memfault overlay config (:file:`overlay-memfault.conf`) includes the most common configuration
+options for using Memfault with modem traces.
+
+To build the MoSh sample with nRF9160 DK and modem traces with flash backend, use the ``-DDTC_OVERLAY_FILE=nrf9160dk_ext_flash.overlay`` and  ``-DOVERLAY_CONFIG="overlay-modem-trace.conf;overlay-memfault.conf"`` options.
```
[the remark this answers is not in the document]
Comment (divipillai) [reply samples/nrf9160/modem_shell/README.rst]: This sentence is now moved to "nRF9160 DK and modem trace support" section under "Building and running".

File: samples/nrf9160/modem_shell/README.rst [hunk 20]
```diff
@@ -453,6 +453,87 @@ Examples
...
+Modem traces
+============
+
+MoSh command: ``modem_trace``
+
+You can use the modem trace commands to control the trace functionality in the modem.
+See :ref:`modem_trace_module` for more information on how to configure modem tracing and the built-in trace backends available.
+
+You need a trace backend that can store modem traces if you want to upload modem traces to the cloud.
+The flash backend can store modem traces to the external flash on the nRF9160 DK and can be  retrieved for uploading.
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/README.rst:468]:
```suggestion
The flash backend can store modem traces to the external flash on the nRF9160 DK and can be retrieved for uploading.
```

File: samples/nrf9160/modem_shell/README.rst [hunk 21]
```diff
@@ -453,6 +453,87 @@ Examples
...
+
+See the following figure, which shows how to download the modem trace data in the `Memfault`_:
+
+.. figure:: /images/modem_shell_trace_download.png
+   :alt: Modem trace download
+
+   Modem trace download
+
+.. note::
+   The conversion of modem trace file to a Wireshark-compatible mode will be available in one of the future releases of the Trace Collector tool.
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/README.rst:501]: mode -> format?

This would sound better to me but I don't know if it changes you intended meaning so up to you to decide.

File: samples/nrf9160/modem_shell/README.rst [hunk 22]
```diff
@@ -1267,6 +1348,19 @@ Use the following command to disconnect from the LwM2M server:
 
 When connected, the ``location`` and ``gnss`` commands use the LwM2M cloud connection for fetching GNSS assistance data and for cellular positioning.
 
+.. _modem_shell_trace_supprot:
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/README.rst:1351]:
```suggestion
.. _modem_shell_trace_support:
```

File: samples/nrf9160/modem_shell/README.rst [hunk 23]
```diff
@@ -453,6 +453,87 @@ Examples
...
+
+.. figure:: /images/modem_shell_trace_download.png
+   :alt: Modem trace download
+
+   Modem trace download
+
+.. note::
+   The conversion of modem trace file to a Wireshark-compatible mode will be available in one of the future releases of the Trace Collector tool.
+
+To build the MoSh sample with the nRF9160 DK and modem traces with a flash backend, see :ref:`modem_shell_trace_supprot`.
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/README.rst:503]:
```suggestion
To build the MoSh sample with the nRF9160 DK and modem traces with a flash backend, see :ref:`modem_shell_trace_support`.
```

**Review by trantanen — APPROVED** [review]
Few small things but don't let them hold this PR.

File: samples/nrf9160/modem_shell/README.rst (hunk 21 above)
Comment (divipillai) [reply samples/nrf9160/modem_shell/README.rst:501]: I think format would sound better here.

**Review by tokangas — APPROVED** [review]

**Review by jhirsi — APPROVED** [review]

**Review by jhn-nordic — APPROVED** [review]

**Review by jtguggedal — APPROVED** [review]

**Review by divipillai — APPROVED** [review]

**Review by eivindj-nordic — APPROVED** [review]
